### PR TITLE
RDM-4885 Validation for the [CREATOR] role in AuthorisationCaseField …

### PR DESCRIPTION
**Pull request by rafalkalita**
…and AuthorisationComplexType

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-4885

### Change description ###
Now if we use [CREATOR] caseRole in AuthorisationCaseField or AuthorisationComplexType it has to be defined in the CaseRoles as well or the import will fail.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ X] Yes
[ ] No
```

### CODE REVIEW

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityACLValidatorImpl.java [hunk 1]
```diff
@@ -14,12 +21,15 @@ public ValidationResult validate(final CaseFieldEntity caseField,
                                      final CaseFieldEntityValidationContext caseFieldEntityValidationContext) {
 
         final ValidationResult validationResult = new ValidationResult();
-
+        Set<String> caseRoles = caseFieldEntityValidationContext.getCaseRoles().stream().map(CaseRoleEntity::getReference).collect(Collectors.toSet());
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityACLValidatorImpl.java]: we could have this:
`.stream().map(CaseRoleEntity::getReference).collect(Collectors.toSet())` inside the `getCaseRoles()` method itself. Most of the times you'll only be interested on the reference

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityValidationContext.java [hunk 2]
```diff
@@ -29,4 +33,7 @@ public String getCaseReference() {
         return this.caseReference;
     }
 
+    public List<CaseRoleEntity> getCaseRoles() {
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityValidationContext.java]: we should rename this to:
`getCaseTypeCaseRoles` 
it's a bit confusing otherwise

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityACLValidatorImpl.java [hunk 3]
```diff
@@ -14,12 +21,15 @@ public ValidationResult validate(final CaseFieldEntity caseField,
...
         for (CaseFieldACLEntity entity : caseField.getCaseFieldACLEntities()) {
 
             if (null == entity.getUserRole()) {
                 validationResult.addError(new CaseFieldEntityInvalidUserRoleValidationError(entity,
                     new AuthorisationCaseFieldValidationContext(caseField, caseFieldEntityValidationContext)));
+            } else if (isCaseRole(entity.getUserRole().getReference()) && !caseRoles.contains(entity.getUserRole().getReference())) {
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityACLValidatorImpl.java]: does this mean that [CREATOR] needs to be declared as a case role in the CaseRoles spreadsheet tab now?

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityACLValidatorImpl.java [hunk 4]
```diff
@@ -14,12 +21,15 @@ public ValidationResult validate(final CaseFieldEntity caseField,
...
+                validationResult.addError(new CaseFieldEntityInvalidUserRoleValidationError(entity,
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityACLValidatorImpl.java]: beware we are throwing an `CaseFieldEntityInvalidUserRoleValidationError` here. So the error message would be: 
`Invalid UserRole for case type '%s', case field '%s'"` which I'm sure will confuse people. Because case roles have not the same syntax of user roles. 
We should create a new error class, with a dedicated error message. Error reporting is very important. It should say something like:
`Unknown case role <x>`. Please make sure <x> is declared in the list of supported case roles for the case type`

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityComplexFieldACLValidatorImpl.java [hunk 5]
```diff
@@ -1,20 +1,25 @@
...
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
 @Component
 public class CaseFieldEntityComplexFieldACLValidatorImpl implements CaseFieldEntityValidator {
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityComplexFieldACLValidatorImpl.java]: same considerations as above apply

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/CaseRoleServiceImpl.java [hunk 6]
```diff
@@ -42,4 +42,11 @@ public CaseRoleServiceImpl(CaseRoleRepository caseRoleRepository,
             .map(dtoMapper::map)
             .collect(toList());
     }
+
+    public static boolean isCaseRole(final String reference) {
+        return reference != null
+            && reference.length() > 2
```
[the remark this answers is not in the document]
Comment (rafalkalita) [reply domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/CaseRoleServiceImpl.java]: I think it is not worth it. Will add more test scenarios covering null and empty.

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityACLValidatorImpl.java (hunk 3 above)
Comment (rafalkalita) [reply domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityACLValidatorImpl.java]: Yes, that is correct.

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityInvalidCaseRoleValidationError.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+public class CaseFieldEntityInvalidCaseRoleValidationError extends ValidationError {
+
+    private final Authorisation caseFieldACLEntity;
+    private final AuthorisationCaseFieldValidationContext authorisationCaseFieldValidationContext;
+
+    public CaseFieldEntityInvalidCaseRoleValidationError(final Authorisation entity,
+                                                         final AuthorisationCaseFieldValidationContext context) {
+        super(String.format("Unknown case role '%s'. Please make sure it is declared in the list of supported case roles for the case type '%s'",
+            context.getCaseReference(),
+            context.getCaseFieldReference()));
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityInvalidCaseRoleValidationError.java]: is this correct?

File: domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityComplexFieldACLValidatorImplTest.java [hunk 8]
```diff
@@ -124,6 +139,28 @@ public void shouldHaveValidationErrorWhenUserNotFound() {
...
+        fieldACL.setUserRole(userRole);
+        caseField.addCaseFieldACL(fieldACL);
+
+        final ValidationResult result = validator.validate(caseField, caseFieldEntityValidationContext);
+
+        assertAll(
+            () -> assertThat(result.getValidationErrors().size(), is(2)),
+            () -> assertThat(result.getValidationErrors().get(0), instanceOf(CaseFieldEntityInvalidCaseRoleValidationError.class)),
+            () -> assertThat(result.getValidationErrors().get(0).getDefaultMessage(), is(
+                "Unknown case role 'case_type'. Please make sure it is declared in the list of supported case roles for the case type 'case_field'"))
```
Comment (mario-paniccia) [inline domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/validation/casefield/CaseFieldEntityComplexFieldACLValidatorImplTest.java]: is this correct? why are we using `case_type` and `case_field`? the first one should be the case role and the second one the case type.

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/endpoint/exception/SpreadsheetValidationErrorMessageCreatorTest.java [hunk 9]
```diff
@@ -425,6 +425,48 @@ public void caseFieldUserRoleEntityInvalidUserRole_createErrorMessageCalled_defa
...
+            .thenReturn(
+                definitionDataItem(SheetName.AUTHORISATION_CASE_FIELD,
+                    new ImmutablePair<>(ColumnName.CASE_TYPE_ID, "case type"),
+                    new ImmutablePair<>(ColumnName.CASE_FIELD_ID, "case field"),
+                    new ImmutablePair<>(ColumnName.USER_ROLE, "X"),
+                    new ImmutablePair<>(ColumnName.CRUD, "Y"))
+            );
+
+        assertEquals(
+            "Invalid case role 'X' in AuthorisationCaseField tab, case type 'case type', case field 'case field', crud 'Y'." +
```
Comment (mario-paniccia) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/endpoint/exception/SpreadsheetValidationErrorMessageCreatorTest.java]: minor:
we should leave CRUD out. Not really relevant in this case
